#Εργασία 1: **Παλινδρόμηση στο Abalone Dataset**


> *Για την παρακάτω εργασια εχω δουλεψει σε Google Colab.*



In [1]:
#Import the basic packages (we will import some more below)
from sklearn import datasets # Library for machine learning
import numpy as np # Library for math and arrays
np.set_printoptions(precision=3) # Tell numpy to print float numbers with 3-decimal precision
import matplotlib.pyplot as plt # Library for graphs and plots
from IPython.display import display, Markdown

#**1. Άνοιγμα αρχείου και προεργασία**

##Φόρτωση της Abalone Data Set



In [2]:
#It will prompt you to select a file. Select and upload the file "abalone.data".
#You should be able to see the name of the file once Colab has uploaded it.
from google.colab import files
uploaded = files.upload()

#χρησιμοποιούμε τη βιβλιοθήκη csv για να χωρίσουμε τα στοιχεία του αρχείου 
#και να δημιουργήσουμε έτσι μια λίστα, κάθε στοιχείο της οποίας είναι επίσης λίστα, και αντιστοιχεί σε μια γραμμή/αντικείμενο
import csv
results = []
with open('abalone.data') as myfile:
  data = csv.reader(myfile, delimiter=',')
  for dataline in data:
    results.append(dataline)
#print(results)
 



Saving abalone.data to abalone (2).data


## Δημιουργία Χ και Υ πινάκων χρησιμοποιωντας slicing


In [3]:
myArray = np.array(results)
Abalone_X, Abalone_Y = myArray[:, 1:8], myArray[:, np.newaxis, -1]

Abalone_X = Abalone_X.astype(float)
Abalone_Y = Abalone_Y.astype(float)

import pandas as pd
df = pd.DataFrame(data= np.c_[Abalone_X , Abalone_Y], columns=['Length', 'Diameter', 'Height', 'Whole weight', 'Shucked weight', 'Viscera weight', 'Shell weight', 'Target'])
df.head(15)


Length  Diameter  Height  Whole weight  Shucked weight  Viscera weight  \
0    0.455     0.365   0.095        0.5140          0.2245          0.1010   
1    0.350     0.265   0.090        0.2255          0.0995          0.0485   
2    0.530     0.420   0.135        0.6770          0.2565          0.1415   
3    0.440     0.365   0.125        0.5160          0.2155          0.1140   
4    0.330     0.255   0.080        0.2050          0.0895          0.0395   
5    0.425     0.300   0.095        0.3515          0.1410          0.0775   
6    0.530     0.415   0.150        0.7775          0.2370          0.1415   
7    0.545     0.425   0.125        0.7680          0.2940          0.1495   
8    0.475     0.370   0.125        0.5095          0.2165          0.1125   
9    0.550     0.440   0.150        0.8945          0.3145          0.1510   
10   0.525     0.380   0.140        0.6065          0.1940          0.1475   
11   0.430     0.350   0.110        0.4060          0.1675          0.0810   
12   0.490     0.380   0.135        0.5415          0.2175          0.0950   
13   0.535     0.405   0.145        0.6845          0.2725          0.1710   
14   0.470     0.355   0.100        0.4755          0.1675          0.0805   

    Shell weight  Target  
0          0.150    15.0  
1          0.070     7.0  
2          0.210     9.0  
3          0.155    10.0  
4          0.055     7.0  
5          0.120     8.0  
6          0.330    20.0  
7          0.260    16.0  
8          0.165     9.0  
9          0.320    19.0  
10         0.210    14.0  
11         0.135    10.0  
12         0.190    11.0  
13         0.205    10.0  
14         0.185    10.0

## Παρουσίαση των χαρακτηριστικων του dataset όπως προκύπτουν από τους πίνακες X και Υ:
Πλήθος αντικειμένων, πλήθος μεταβλητών, εύρος τιμών του y




In [4]:
display(Markdown(
    "**Πλήθος αντικειμένων, Πλήθος μεταβλητών του Χ**: " + str(np.shape(Abalone_X)) + "<br>" +
    "**Πλήθος αντικειμένων, Πλήθος μεταβλητών του Υ**: "+ str(np.shape(Abalone_Y)) + "<br>" +
    "**Εύρος τιμών του Υ** : " + str(np.max(Abalone_Y) - np.min(Abalone_Y))
    ))

**Πλήθος αντικειμένων, Πλήθος μεταβλητών του Χ**: (4177, 7)<br>**Πλήθος αντικειμένων, Πλήθος μεταβλητών του Υ**: (4177, 1)<br>**Εύρος τιμών του Υ** : 28.0

## Εκτύπωση του μέσου και της τυπικής απόκλισης της κάθε μεταβλητής.

In [5]:
display(Markdown(
    "**Το Μέσο της κάθε μεταβλητης από τον πίνακα X** : " + str(np.mean(Abalone_X, axis=0))  + "<br>" +
    "**Το Μέσο της κάθε μεταβλητης από τον πίνακα Y** : " + str(np.mean(Abalone_Y, axis=0))  + "<br>" +
    "**Η Τυπική Απόκλιση της κάθε μεταβλητης από τον πίνακα Χ** : " + str(np.std(Abalone_X,0))  + "<br>" +
    "**Η Τυπική Απόκλιση της κάθε μεταβλητης από τον πίνακα Y** : " + str(np.std(Abalone_Y,0))
))

**Το Μέσο της κάθε μεταβλητης από τον πίνακα X** : [0.524 0.408 0.14  0.829 0.359 0.181 0.239]<br>**Το Μέσο της κάθε μεταβλητης από τον πίνακα Y** : [9.934]<br>**Η Τυπική Απόκλιση της κάθε μεταβλητης από τον πίνακα Χ** : [0.12  0.099 0.042 0.49  0.222 0.11  0.139]<br>**Η Τυπική Απόκλιση της κάθε μεταβλητης από τον πίνακα Y** : [3.224]

##Χωρίζω τα δεδομένα σε training set και test set με seed=5

In [6]:
# Split the data randomly (stratified) with a random seed
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(
    Abalone_X, Abalone_Y, test_size=0.1, random_state=5)



##***Normalization***

*The goal of normalization is to to make all the elements lie between 0 and 1 thus bringing all the values of numeric columns in the dataset to a common scale. The Abalone Data Set in my opinion does not require normalization because the mean of the training set is already between 0 and 1.*

In [7]:
print(np.mean(X_train, axis=0))

[0.524 0.408 0.14  0.83  0.36  0.181 0.239]


# **2. Κριτήριο αξιολόγησης αποτελεσμάτων**
##Συνάρτηση: **Success_Rate**

In [8]:
def success_rate(y_test_pred, y_test, N):
  successes=0
  a = np.absolute(y_test_pred - y_test)
  for x in a:
    if (np.less_equal(x, N)):
      successes = successes + 1

  return (successes/np.count_nonzero(a)) * 100

# **3. Γραμμικά μοντέλα**

## Εκπαίδευση μοντέλου πολλαπλής παλινδρόμησης

In [9]:
from sklearn import linear_model
from sklearn.metrics import explained_variance_score

# Create linear regression object
regr1 = linear_model.LinearRegression()

# Train the model using the training sets
regr1.fit(X_train, y_train)
Abalone_y_pred = regr1.predict(X_test)

## Εκτύπωση αποτελεσμάτων πολλαπλής παλινδρόμησης

In [10]:
display(Markdown(
    "**Test set explained variance**: {:.2f}".format(explained_variance_score(y_test, Abalone_y_pred))  + "<br>" +
    "**Success Rate**:  {:.2f}".format(success_rate(Abalone_y_pred, y_test, 1))
))

**Test set explained variance**: 0.48<br>**Success Rate**:  40.67

## **Ridge Regularization**

In [11]:
from sklearn.model_selection import GridSearchCV

params = { 
           'alpha': np.logspace(-6, 4, 20) # try 20 values in logarithmic scale from 10^-6 to 10^4
         }
print("alphas: ")
print(params['alpha'])

regr2 = linear_model.Ridge() 
grid = GridSearchCV(estimator=regr2, cv = 10, param_grid=params) 
grid.fit(X_train, y_train) 
y_train_pred = grid.predict(X_train)
y_test_pred = grid.predict(X_test)

alphas: 
[1.000e-06 3.360e-06 1.129e-05 3.793e-05 1.274e-04 4.281e-04 1.438e-03
 4.833e-03 1.624e-02 5.456e-02 1.833e-01 6.158e-01 2.069e+00 6.952e+00
 2.336e+01 7.848e+01 2.637e+02 8.859e+02 2.976e+03 1.000e+04]


## Εκτύπωση αποτελεσμάτων regularized πολλαπλής παλινδρόμησης

In [12]:
display(Markdown(
    "**Test set explained variance**: {:.2f}".format(explained_variance_score(y_test, y_test_pred))  + "<br>" +
    "**Success Rate**:  {:.2f}".format(success_rate(y_test_pred, y_test, 1))
))

**Test set explained variance**: 0.47<br>**Success Rate**:  39.71

##Εκτύπωση 𝐰 συντελεστών για την πολλαπλή γραμμική παλινδρόμηση και για την regularized πολλαπλή παλινδρόμηση.

In [13]:
display(Markdown(
    "### Πολλαπλή Γραμμική Παλινδρόμηση:\n" +
    "**Coefficients**: " + str(regr1.coef_)
))

display(Markdown(
    "### Regularized Πολλαπλή Παλινδρόμηση: \n" +
    "**Coefficients**: " + str(grid.best_estimator_.coef_)
))

### Πολλαπλή Γραμμική Παλινδρόμηση:
**Coefficients**: [[ -0.597  12.417  10.648   9.04  -19.784 -10.076   9.177]]

### Regularized Πολλαπλή Παλινδρόμηση: 
**Coefficients**: [[  2.174   8.693   8.692   7.641 -18.18   -7.615  10.518]]

# **4. Μη-γραμμικά μοντέλα**

##Πολυωνυμική παλινδρόμηση με όλες τις μεταβλητές (2ου βαθμού) 


In [14]:
from sklearn.preprocessing import PolynomialFeatures

#Create a 2nd degree polynomial model upon which to fit the data
polyModel = PolynomialFeatures(degree = 2)
X_train_poly = polyModel.fit_transform(X_train)
X_test_poly = polyModel.fit_transform(X_test)
print("Training data shape:", np.shape(X_train_poly))
print("Test data shape:", np.shape(X_test_poly))

Training data shape: (3759, 36)
Test data shape: (418, 36)


## **Ridge Regularization**

In [15]:
regr = linear_model.Ridge() 
grid = GridSearchCV(estimator=regr, cv = 10, param_grid=params) 
grid.fit(X_train_poly, y_train) 
y_train_pred2 = grid.predict(X_train_poly)
y_test_pred2 = grid.predict(X_test_poly)

##Συγκρίση των αποτελέσματων με αυτά των γραμμικών μοντέλων

In [16]:
display(Markdown(
    "### Regularized Πολυωνυμική Παλινδρόμηση:\n" +
    "**Test set explained variance**: {:.2f}".format(explained_variance_score(y_test, y_test_pred2))  + "<br>" +
    "**Success Rate**:  {:.2f}".format(success_rate(y_test_pred2, y_test, 1))
))

display(Markdown(
    "### Regularized Πολλαπλή Παλινδρόμηση:\n" +
    "**Test set explained variance**: {:.2f}".format(explained_variance_score(y_test, y_test_pred))  + "<br>" +
    "**Success Rate**:  {:.2f}".format(success_rate(y_test_pred, y_test, 1))
))

### Regularized Πολυωνυμική Παλινδρόμηση:
**Test set explained variance**: 0.53<br>**Success Rate**:  43.06

### Regularized Πολλαπλή Παλινδρόμηση:
**Test set explained variance**: 0.47<br>**Success Rate**:  39.71

####Κατα την γνώμη μου υπήρξε βελτίωση καθώς το Success Rate απο 39.71 που ήταν στο γραμμικό μοντέλο αυξήθηκε στο 43.06 στο πολυωνυμικό.

#k Nearest Neighbours Regression.

In [17]:
from sklearn.neighbors import KNeighborsRegressor
regr = KNeighborsRegressor()
params = { 
           'n_neighbors': range(1,10),
            'weights': ['uniform', 'distance']
         }
grid = GridSearchCV(estimator=regr, cv = 10, param_grid=params) 
grid.fit(X_train, y_train) 
print(grid.best_score_)
y_test_pred = grid.predict(X_test)
display(Markdown(
    "**k-Nearest Neighbours Regression**: <br>" +
    "**Best k**: {:.3f}".format(grid.best_params_['n_neighbors'])  + "<br>" +
    "**Best weighting strategy**: " + grid.best_params_['weights']  + "<br>" +
    "**Success Rate**:  {:.2f}".format(success_rate(y_test_pred, y_test, 1)) + "<br>" +
    "**Test set explained variance**: {:.2f}".format(explained_variance_score(y_test, y_test_pred))
))

0.5340870108664669


**k-Nearest Neighbours Regression**: <br>**Best k**: 9.000<br>**Best weighting strategy**: distance<br>**Success Rate**:  45.45<br>**Test set explained variance**: 0.53

####Τα αποτελέσματα είναι καλύτερα καθώς το Success Rate αυξηθηκε στο 45.45.